header-validator/src/maybe.ts

```diff
@@ -1,29 +1,30 @@
-export type Maybe<T> = typeof None | Some<T>
+export type Maybeable<T> = T | Maybe<T>
 
-export const None = {
-  filter: () => None,
-  map: () => None,
-  flatMap: () => None,
-  peek: () => None,
-}
+export class Maybe<T> {
+  static readonly None: Maybe<any> = new Maybe()
 
-export class Some<T> {
-  constructor(private readonly t: T) {}
+  static some<T>(t: T): Maybe<T> {
+    return new Maybe(t)
+  }
 
-  filter(f: (t: T) => boolean): Maybe<T> {
-    return f(this.t) ? this : None
+  static flatten<T>(t: Maybeable<T>): Maybe<T> {
+    return t instanceof Maybe ? t : Maybe.some(t)
   }
 
-  map<U>(f: (t: T) => U): Some<U> {
-    return new Some(f(this.t))
+  private constructor(private readonly t?: T) {}
+
+  filter(f: (t: T) => boolean): Maybe<T> {
+    return this.t === undefined || !f(this.t) ? Maybe.None : this
   }
 
-  flatMap<U>(f: (t: T) => Maybe<U>): Maybe<U> {
-    return f(this.t)
+  map<U>(f: (t: T) => Maybeable<U>): Maybe<U> {
```

Comment:
```suggestion
  map<U>(f: (t: T) => U | Maybe<U>): Maybe<U> {
```

I feel like having the type inline here might be simpler if it is its only use.

Reply:
It will be used in the followup PR, but noted.